### PR TITLE
Add production environment

### PR DESCRIPTION
**Pull request by benjaminbrassart**
Changes:
- Use Dockerfiles instead of writing the commands inside docker-compose.yml
- Move `.env` to `.env.development`
- Add `.env.production`
- Add script `setenv.sh` to switch between environments
- Add docker-compose.dev.yml (dev)
- Add docker-compose.override.yml (prod)
- Change Makefile so it uses the development docker-compose configuration
- Update readme with setup instructions

### CODE REVIEW

**Comment by benjaminbrassart** [comment]
https://stackoverflow.com/a/52430444

File: backend/src/main.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ async function bootstrap() {
```
Comment (Balgor18) [inline backend/src/main.ts:27]: Question about the line 24. Do we really need a adapter for the socket ?

File: docker-compose.yml [hunk 2]
```diff
@@ -41,56 +27,38 @@ services:
...
-      INTRA42_CALLBACK_URL: ${INTRA42_CALLBACK_URL}
-      GITHUB_CLIENT_ID: ${GITHUB_CLIENT_ID}
-      GITHUB_CLIENT_SECRET: ${GITHUB_CLIENT_SECRET}
-      GITHUB_CALLBACK_URL: ${GITHUB_CALLBACK_URL}
-    networks:
-      - transcendence
-    volumes:
-      - ./backend:/backend
-      - backend_node_modules:/backend/node_modules/
+      NEST_DEBUG: true # TODO REMOVE THIS ('true' break compilation on other machines)
```
Comment (Balgor18) [inline docker-compose.yml:33]: We dont need it anymore. We can erase

**Review by Balgor18 — CHANGES_REQUESTED** [review]
i just add 2 comment about remove 1 things for sure and the other i think we dont need it

File: frontend/.env [hunk 3]
```diff
@@ -0,0 +1 @@
+../.env
```
Comment (BarbaraC12) [inline frontend/.env:1]: is this very necessary ?

**Review by BarbaraC12 — APPROVED** [review]

**Comment by jesuisstan** [comment]
Cannot login neigher with 42 nor with Github. Guess, something is wrong with CALLBACK_URL

in address line see this: "http://localhost:3080/auth/github/callback?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-authorization-request-errors%2F%23redirect-uri-mismatch"

File: docker-compose.yml (hunk 2 above)
Comment (benjaminbrassart) [reply docker-compose.yml:33]: Done

File: frontend/.env (hunk 3 above)
Comment (benjaminbrassart) [reply frontend/.env:1]: The file is a symbolic link, its content is the link's target. I suppose, however, that we could get rid of it

**Comment by benjaminbrassart** [comment]
> Cannot login neigher with 42 nor with Github. Guess, something is wrong with CALLBACK_URL
> 
> in address line see this: "http://localhost:3080/auth/github/callback?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-authorization-request-errors%2F%23redirect-uri-mismatch"

Backend url becomes `{frontendUrl}/api` in production, the backend is hidden behind a reverse proxy. This is because CORS errors are raised when using http on another host than localhost. So, the redirect url for github oauth flow would be `http://10.2.1.5:3000/api/auth/github/redirect`

**Review by Balgor18 — APPROVED** [review]

File: frontend/.env (hunk 3 above)
Comment (benjaminbrassart) [reply frontend/.env]: Done

File: backend/src/main.ts (hunk 1 above)
Comment (benjaminbrassart) [reply backend/src/main.ts:27]: I honestly don't know, I merged with what had been done previously.

**Review by Balgor18 — APPROVED** [review]

**Review by BarbaraC12 — APPROVED** [review]

**Review by BarbaraC12 — APPROVED** [review]

**Review by Balgor18 — APPROVED** [review]

**Review by Balgor18 — APPROVED** [review]

**Review by jesuisstan — COMMENTED** [review]
ok

**Review by jesuisstan — APPROVED** [review]

**Pull request merged**